libcudacxx/include/cuda/std/detail/libcxx/include/optional

```diff
@@ -1397,26 +1457,312 @@ operator>=(const _Tp& __v, const optional<_Up>& __x)
+  {}
+
+  _CCCL_TEMPLATE(class _Up)
+  _CCCL_REQUIRES(__from_temporary<const _Up&>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr optional(const optional<_Up>&) = delete;
+
+  _CCCL_TEMPLATE(class _Up)
+  _CCCL_REQUIRES(_CCCL_TRAIT(is_convertible, _Up, _Tp&) _CCCL_AND(!__from_temporary<_Up>))
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr optional(optional<_Up>&& __u) noexcept(
+    noexcept(static_cast<_Tp&>(_CUDA_VSTD::declval<_Up>())))
+      : __value_(
+          __u.has_value() ? _CUDA_VSTD::addressof(static_cast<_Tp&>(_CUDA_VSTD::forward<_Up>(__u.value()))) : nullptr)
+  {}
+
+  _CCCL_TEMPLATE(class _Up)
+  _CCCL_REQUIRES((!_CCCL_TRAIT(is_convertible, _Up, _Tp&)) _CCCL_AND _CCCL_TRAIT(is_constructible, _Tp&, _Up)
+                   _CCCL_AND(!__from_temporary<_Up>))
+  _LIBCUDACXX_HIDE_FROM_ABI explicit constexpr optional(optional<_Up>&& __u) noexcept(
+    noexcept(static_cast<_Tp&>(_CUDA_VSTD::declval<_Up>())))
+      : __value_(
+          __u.has_value() ? _CUDA_VSTD::addressof(static_cast<_Tp&>(_CUDA_VSTD::forward<_Up>(__u.value()))) : nullptr)
+  {}
+
+  _CCCL_TEMPLATE(class _Up)
+  _CCCL_REQUIRES(__from_temporary<_Up>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr optional(optional<_Up>&&) = delete;
+
+  _CCCL_TEMPLATE(class _Up)
+  _CCCL_REQUIRES(_CCCL_TRAIT(is_convertible, const _Up, _Tp&) _CCCL_AND(!__from_temporary<const _Up>))
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr optional(const optional<_Up>&& __u) noexcept(
+    noexcept(static_cast<_Tp&>(_CUDA_VSTD::declval<const _Up>())))
+      : __value_(__u.has_value() ? _CUDA_VSTD::addressof(static_cast<_Tp&>(__u.value())) : nullptr)
+  {}
+
+  _CCCL_TEMPLATE(class _Up)
+  _CCCL_REQUIRES((!_CCCL_TRAIT(is_convertible, const _Up, _Tp&)) _CCCL_AND _CCCL_TRAIT(
+    is_constructible, _Tp&, const _Up) _CCCL_AND(!__from_temporary<const _Up>))
+  _LIBCUDACXX_HIDE_FROM_ABI explicit constexpr optional(const optional<_Up>&& __u) noexcept(
+    noexcept(static_cast<_Tp&>(_CUDA_VSTD::declval<const _Up>())))
+      : __value_(__u.has_value() ? _CUDA_VSTD::addressof(static_cast<_Tp&>(__u.value())) : nullptr)
+  {}
+
+  _CCCL_TEMPLATE(class _Up)
+  _CCCL_REQUIRES(__from_temporary<const _Up>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr optional(const optional<_Up>&&) = delete;
+
+  _CCCL_HIDE_FROM_ABI constexpr optional& operator=(const optional&) noexcept = default;
+  _CCCL_HIDE_FROM_ABI constexpr optional& operator=(optional&&) noexcept      = default;
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr optional& operator=(nullopt_t) noexcept
+  {
+    __value_ = nullptr;
+    return *this;
+  }
+
+  _CCCL_TEMPLATE(class _Up = _Tp)
+  _CCCL_REQUIRES((!__is_std_optional<decay_t<_Up>>::value) _CCCL_AND _CCCL_TRAIT(is_constructible, _Tp&, _Up))
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr optional&
+  operator=(_Up&& __u) noexcept(noexcept(static_cast<_Tp&>(_CUDA_VSTD::forward<_Up>(__u))))
+  {
+    __value_ = _CUDA_VSTD::addressof(__u);
+    return *this;
+  }
```

Comment:
I don't see this overload in the paper, just a defaulted assignment from `const optional&`. Are you trying to shortcut a converting constructor here? If so, this doesn't look right (for one, it will accept things that require an explicit constructor call otherwise; it also lacks a check for assignment from a temporary). I think this should be stricken.

Reply:
That's in the implementation section, but it is no longer there in the wording section.e

Reply:
Seems like I looked at the reference implementation and not the wording ^^

Dropped